dev/PackageManager/API/PackageManager.idl

```diff
@@ -0,0 +1,407 @@
+        Windows.ApplicationModel.PackageVersion MinVersion; //required
+        Microsoft.Windows.ApplicationModel.DynamicDependency.PackageDependencyProcessorArchitectures ProcessorArchitectureFilter;
+
+        // Source where a package can be retrieved from if/when needed for install, remediation, etc
+        Windows.Foundation.Uri PackageUri;
+
+        // ??? Additional info potentially needed for deployment operations
+        //TODO see AddPackageOptions
+
+        // Additional info for runtime use
+        //TODO see CreatePackageDependencyOptions
+        //TODO see AddPackageDependencyOptions
+    }
+
+    [feature(Feature_PackageManager)]
+    [contract(PackageDeploymentContract, 1)]
+    runtimeclass PackageSet
+    {
+        PackageSet();
+
+        // Unique id
+        String Id;
+
+        // Package(s) in the set
+        IVector<PackageSetItem> PackageSetItems { get; };
+    }
+
+    [feature(Feature_PackageManager)]
+    [contract(PackageDeploymentContract, 1)]
+    runtimeclass FindPackageOptions
+    {
+        FindPackageOptions();
+
+        //TODO
+        Boolean TODO_MustHaveAtLeastOneInterfaceOrStaticFactoryPlaceholder;
+    }
+
+    [feature(Feature_PackageManager)]
+    [contract(PackageDeploymentContract, 1)]
+    runtimeclass PackageSetManager
+    {
+        static PackageSetManager GetDefault();
+
+        void Add(PackageSet packageSet);  		                // Create/Write/Save
+        void Update(PackageSet packageSet);  		            // Update/Write/Save
+        void Remove(String packageSetId);		                // Delete
+        PackageSet Get(String packageSetId);		            // Read/Load
+        IVector<PackageSet> Find(FindPackageOptions options);   // Enumerate
+    };
+
+    [feature(Feature_PackageManager)]
+    [contract(PackageDeploymentContract, 1)]
+    enum PackageDeploymentStatus
+    {
+        InProgress = 0,         // The request is in progress
+        CompletedSuccess = 1,   // The request completed successfully
+        CompletedFailure = 2,   // The request failed with some critical internal error. //TODO:Check ExtendedError for HRESULT
+    };
+
+    [feature(Feature_PackageManager)]
+    [contract(PackageDeploymentContract, 1)]
+    enum PackageDeploymentProgressStatus
+    {
+        Queued = 0,             // The request is queued
+        InProgress = 1,         // The request is in progress
+        CompletedSuccess = 2,   // The request completed successfully
+        CompletedFailure = 3,   // The request failed with some critical internal error. //TODO:Check ExtendedError for HRESULT
+    };
+
+    [feature(Feature_PackageManager)]
+    [contract(PackageDeploymentContract, 1)]
+    runtimeclass PackageDeploymentProgress
+    {
+        PackageDeploymentProgressStatus status;
+        Double percentage;
+    }
+
+    [feature(Feature_PackageManager)]
+    [contract(PackageDeploymentContract, 1)]
+    runtimeclass PackageDeploymentResult
```

Comment:
>PackageDeploymentResult [](http://example.com/codeflow?start=17&length=23)

This all seems to refer to a singular package whereas all the APIs that return it are for Sets. Would generally worry about trying to build a batched API on top of a windows API that only supports one package per operation. Seems like ideally the c++\winrt concurrency apis would just be straightforward enough to use that making an installing a vector of packages would be straightforward for callers to do themselves, but maybe they aren't.

Reply:
ByDesign. Yes, perhaps, in the future there'll be a good reason to add more-granular-results info.

Maybe. It's not very meaningful today.

The complex data about any problems with a deployment operation are logged (extensively) in the eventlog.

And there's precedent NOT to do that ==> `pkgmgr.AddPackageAsync("file://c:/foo/bar.msixbundle")` can install multiple packages (e.g. Main/x64 + Resource/en-us + Resource/scale-150 + ...), and possibly even fail during the batch operation (e.g. Main/x64 succeeds but Resource/en-us fails) and yet there's only a singular result about the operation, not granular failure details (Main/x64=success, Resource/en-us=fail, Resource/scale-150=not-even-tried, ...).

It's an interesting idea for a future enhancement, perhaps. It hasn't been a requirement to date, nor an impediment or pain point. Hence the reason it's not in the design.

>installing a vector of packages would be straightforward for callers to do themselves, but maybe they aren't.

That's one of those "The only difference between theory and practice is, in theory, there's no difference between theory and practice, whereas in practice..."

Simple concept, but the execution quickly gets complicated. One of the benefits of `PackageDeploymentManager` and `PackageSet`s is _we_, deeply experienced in the space and immersed in the complexities and nuances, can do the heavy lifting to implement such things many would find challenging to do well, for everyone to reap the benefit.

The v0 goal is to unblock early development. This is more of a marathon than a sprint. This is the first PR, not the last :-)